src/main/java/org/spongepowered/api/text/template/TextTemplate.java

```diff
@@ -0,0 +1,244 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.template;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.TextElement;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Represents a re-usable template that produces a formatted
+ * {@link Text.Builder}. Elements will be appended to the result builder in the
+ * order that they are specified in {@link #of(Object...)}.
+ *
+ * TODO: Serialization
+ */
+public class TextTemplate {
+
+    /**
+     * Empty representation of a {@link TextTemplate}. This is returned if the
+     * array supplied to {@link #of(Object...)} is empty.
+     */
+    public static final TextTemplate EMPTY = new TextTemplate();
+
+    final Object[] elements;
+
+    TextTemplate(Object... elements) {
+        this.elements = elements;
+    }
+
+    /**
+     * Returns the elements contained in this TextTemplate.
+     *
+     * @return The elements within the template
+     */
+    public List<Object> getElements() {
+        return Collections.unmodifiableList(Arrays.asList(elements));
+    }
+
+    /**
+     * Evaluates this TextTemplate and returns the result in a
+     * {@link Text.Builder}.
+     *
+     * @param params Parameters to evaluate
+     * @return Text builder containing result
+     * @throws RequiredTextTemplateArgumentException if required parameters are missing
+     */
+    public Text.Builder eval(Map<String, Object> params) {
+        checkNotNull("params");
+        Text.Builder result = Text.builder();
+        for (Object element : this.elements) {
+            appendElement(result, element, params);
+        }
+        return result;
+    }
+
+    /**
+     * Evaluates this TextTemplate and returns the result in a
+     * {@link Text.Builder}.
+     *
+     * @return Text builder containing result
+     * @throws RequiredTextTemplateArgumentException if required parameters are missing
+     */
+    public Text.Builder eval() {
+        return eval(Collections.emptyMap());
+    }
+
+    private void appendElement(Text.Builder builder, Object element, Map<String, Object> params) {
```

Comment:
Along with previous method regarding visibility, this would be nice to have `protected`.